npm/src/compiler/compiler.ts

```diff
@@ -15,15 +15,23 @@ type Wasm = typeof import("../../lib/node/qsc_wasm.cjs");
 // for running the compiler in the same thread the result will be synchronous (a resolved promise).
 export interface ICompiler {
   checkCode(code: string): Promise<VSDiagnostic[]>;
-  getHir(code: string): Promise<string>;
+  getHir(code: string, language_features: string[]): Promise<string>;
   run(
     sources: [string, string][],
     expr: string,
     shots: number,
+    language_features: string[],
```

Comment:
This would be a breaking API change. Could we make this an optional parameter?

Reply:
There is a fairly idiomatic way to make this change backward compatible. You define method overloads in the interface and check the argument type at runtime. See my [example on the TypeScript Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAyg9gWwgeQEYCsIGNgBVzQC8UA3lKgDYCGAFgFxQDOwATgJYB2A5lAL4DcAKEGdgEFgDMqWaAEkAYnDilBUNVAD0AKi1QAAgBMIYFtipiDUMBQhVG0AK72owGtA4QA7lDgA3cRRwVJZaGqrqEkoAFFQsLFQgAIIsXAzM7NwA2gC6AJQMvnBsBkLhapFwUXAY2MDJqbCIKDU4+JAANFCx8UkpaaycXDn5UIXFQrzCWNSMjFCKymwI1hBIHMBzCkoq6lAVMXEJyCxomDj1-RlD2VAAPo1Ip7VtEJ3dCfUA-JeDw6RQZV2UDYEigUWSPQAdGxGBCEgcescnucUrlciRAUD1Fg4BxGHAbJDAlwogAiCRsFjMHz+FiBYIAQlJuSEWKxGg0UBeMCw7DAwCgHG6cE8c1c0FAkBcynSv2ynXsErcUBofnEwI4kR8HC6hxASJadRSmLZOLxBIgRLgJNJjFVDgollQ0FlWWyDFJbz1BrORq4LJNWLN+MJxLJdrgDqdjg4RgpHgMHq9PXqAbZQODFqtJLTbN4UAgFGcGPTu0zoet4ewuMsarpQQMTNzpc0nO5vLY-MFwtFLmVkolyngj0NLwVECV0FVtI1WtxusRJ0N9UDGdxIctYdt9sd5Ggw+avpedE9C6OS99qf4AJbanLm8r28ju+dTAGbqTZ96-tZt-v2aiZsgUmXZJkmQRgwFCoGC2ZRiA8bwFkAoQKkhfZMlJKhT1JVBsKwUk8hQpQ0OiMhKFoD1yJoUk+E6DCsM6HC8IIlkgA) . TypeScript is smart enough to narrow the type of the argument at compile time, and JavaScript provides runtime types so that you can check at runtime which version of the function is being called.

Here, we'd check the type of the fourth argument at runtime, if it's a string array, we know the caller is using the new overload. If not, we know they're calling the old version.

Now, that being said. This will get us through in the short term, but maybe this is a good point to pause and consider our strategy for handling breaking changes in this group of functions going forward.

1. I propose changing the signatures of these methods to something like:

```ts
run(config: ProgramConfig, eventHandler: IQscEventTarget);
getEstimates(config: ProgramConfig, params: string);
..... etc ....

type ProgramConfig  = {
    sources: {
        name: string;
        contents: string;
    }[];
    languageFeatures?: string[];
    // future-proof - we can add other optional properties as time goes on
}
```

IMO, it's more maintainable to make backward-compatible changes to object properties than to function arguments. Optional properties look nicer at the call site (`foo({ bar: "bar"})` looks better than `foo(undefined, undefined, undefined, undefined, "bar")`). Also, since most of these functions take basically the same set of arguments; it makes sense to me to group them together into an object.

2. Requiring the `IQscEventHandler` argument to be the last argument wasn't the greatest decision, and maybe I can change that too while we're at it. I can make it the first argument - gives us a bit more flexibility with future optional arguments.

Reply:
I liked @minestarks's suggestion and the new `run`. I'm in favor of doing this for the rest of the API.

Reply:
@sezna yep that's exactly what I had in mind! I can also take this opportunity to move `IQscEventHandler` to be the first argument? But that may be a bit too much to add to this PR... perhaps it could be a separate change after this PR is merged, but before we ship the next build (since there's no need to have _two_ deprecated signatures if we can avoid it).

Reply:
Sounds fine by me. We can iron out the remaining wrinkles in the follow up PR